sklearn/metric_learning/nca.py

```diff
@@ -0,0 +1,216 @@
+from ..base import BaseEstimator, TransformerMixin
+
+import numpy as np
+import scipy as sp
+import scipy.misc
+import scipy.optimize
+import time
+
+import sys
+
+def nca_vectorized_oracle(X, y, params):
+    dX = X[:, None, :] - X[None, :, :]  # n_samples x n_samples x n_comp
+    outer = dX[:, :, None, :] * dX[:, :, :, None]  # n_samples x n_samples x n_comp x n_comp
+
+    n_samples, n_features = X.shape
+    n_components = params.n_components
+
+    def oracle(L):
+        L = L.reshape((n_components, n_features))
+
+        Lx = np.dot(X, L.T)  # n_samples x n_comp
+        assert Lx.shape == (n_samples, n_components)
+
+        A = Lx[None, :, :] - Lx[:, None, :]  # n_samples x n_samples x n_comp
+        assert A.shape == (n_samples, n_samples, n_components)
+
+        logp = -(A*A).sum(axis=2)  # n_samples x n_samples
+        np.fill_diagonal(logp, -np.inf)
+        logp -= sp.misc.logsumexp(logp, axis=1)[:, None]
+        assert logp.shape == (n_samples, n_samples)
+
+        p = np.exp(logp)  # n_samples x n_samples
+        assert p.shape == (n_samples, n_samples)
+
+        class_neighbours = y[:, None] == y[None, :]
+
+        p_i = (p * class_neighbours).sum(axis=1)
+
+        if params.loss == 'l1':
+            grad = np.sum(np.sum(p[:, :, None, None] * outer, axis=0) * p_i[:, None, None], axis=0)
+        else:
+            grad = np.sum(np.sum(p[:, :, None, None] * outer, axis=0), axis=0)
+
+        assert grad.shape == (n_features, n_features)
+
+        if params.loss == 'l1':
+            snd = np.sum(class_neighbours[:, :, None, None] * p[:, :, None, None] * outer, axis=(0, 1))
```

Comment:
specifying several axes for `np.sum` may not be supported by all numpy versions that sklearn supports. I haven't checked, but it may be necessary to backport or work around this.